packages/core/src/lib/ConnectionManager.ts

```diff
@@ -0,0 +1,336 @@
+     * @see https://github.com/status-im/js-waku/issues/252
+     *
+     * >This event will be triggered anytime we are disconnected from another peer,
+     * >regardless of the circumstances of that disconnection.
+     * >If we happen to have multiple connections to a peer,
+     * >this event will **only** be triggered when the last connection is closed.
+     * @see https://github.com/libp2p/js-libp2p/blob/bad9e8c0ff58d60a78314077720c82ae331cc55b/doc/API.md?plain=1#L2100
+     */
+    this.libp2p.connectionManager.addEventListener("peer:disconnect", (evt) => {
+      this.stopKeepAlive(evt.detail.remotePeer);
+    });
+  }
+
+  private dialPeersInInterval(
+    maxBootstrapPeersAllowed: number,
+    interval: number
+  ): void {
+    // increase interval after every attempt
+    // -> 1st interval: 10 seconds x 2 = 20 seconds
+    // -> 2nd interval: 10 seconds x 3 = 40 seconds
+    // and so on
+    const newInterval = this.dialAttempt * interval;
+    log(`Dialing next set of discovered peers in ${newInterval} ms`);
+
+    // recursively run this function with increase in time
+    this.dialDiscoveredPeers(maxBootstrapPeersAllowed)
+      .then(() => {
+        setTimeout(() => {
+          this.dialPeersInInterval(maxBootstrapPeersAllowed, interval);
+        }, newInterval);
+        this.dialAttempt++;
+      })
+      .catch((err) => {
+        log("Error dialing discovered peers", err);
+      });
+  }
+
+  private async dialDiscoveredPeers(
+    maxBootstrapPeersAllowed: number
+  ): Promise<void> {
+    const {
+      dialableBootstrapPeers,
+      dialableNonBootstrapPeers,
+      bootstrapConnections,
+    } = await this.fetchUpdatedState();
+
+    // find & dial peers found via `dns-discovery`
+    if (
+      (bootstrapConnections.length < maxBootstrapPeersAllowed &&
+        dialableBootstrapPeers.length > 0) ||
+      this.dialAttempt >= DEFAULT_DIAL_ATTEMPTS_BEFORE_BOOTSTRAP_CONNECTION
+    ) {
+      for (let i = 0; i < maxBootstrapPeersAllowed; i++) {
+        const peer = dialableBootstrapPeers[i];
+        if (!peer) break;
+
+        this.dialPeer(peer)
+          .then(() => log(`Dial successful`))
+          .catch((error) => log(error));
+      }
+    }
+
+    //find & dial peers found via discovery mechanisms other than `dns-discovery`
+    for (const peer of dialableNonBootstrapPeers) {
+      this.dialPeer(peer)
+        .then(() => log(`Dial successful`))
+        .catch((error) => log(error));
+    }
+  }
+
+  private async fetchUpdatedState(): Promise<UpdatedStates> {
+    const allPeers = await this.libp2p.peerStore.all();
+
+    const allDialablePeers: Peer[] = [];
+    const bootstrapPeers: Peer[] = [];
+    const nonBootstrapPeers: Peer[] = [];
+    const dialableBootstrapPeers: Peer[] = [];
+    const dialableNonBootstrapPeers: Peer[] = [];
+    for (const peer of allPeers) {
+      const isBootstrap = (await this.libp2p.peerStore.getTags(peer.id)).some(
```

Comment:
I think we can optimise a bit here by avoiding usage of `await` within `for` loop and moving to `Promise.all` approach. Because currently program will process peers one by one and will await for tags each time.